### PR TITLE
Refactor data reader and implement a CSVDataReader.

### PR DESCRIPTION
**Pull request by workingloong**
Move ODPSDataReader and RecordIODataReader to a single file. 
Implement a CSVDataReader to read a CSV file to support locally train a Keras model.

### CODE REVIEW

File: elasticdl/python/data/odps_reader.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
+import tensorflow as tf
```
Comment (brightcoder01) [inline elasticdl/python/data/odps_reader.py]: elasticdl/python/data/odps_reader.py -> elasticdl/python/data/readers/odps_reader.py

Comment (workingloong) [reply elasticdl/python/data/odps_reader.py]: Done

File: elasticdl/python/data/reader/csv_reader.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    def read_records(self, task):
+        with open(task.shard_name, "r", encoding="utf-8") as csv_file:
+            csv_reader = csv.reader(csv_file, delimiter=self.seq)
+            csv_columns = next(csv_reader)
+            selected_columns = (
+                csv_columns
+                if self.selected_columns is None
+                else self.selected_columns
+            )
+            if not set(csv_columns) >= set(selected_columns):
```
Comment (QiJune) [inline elasticdl/python/data/reader/csv_reader.py]: `if not set(csv_columns) >= set(selected_columns):` to

`if set(selected_columns) > set(csv_columns)`

File: elasticdl/python/data/reader/csv_reader.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                raise ValueError(
+                    "The first line in the csv file must be column names and"
+                    "the selected columns is not in the file. The selected"
+                    "columns are {} and the columns in {} are {}".format(
+                        selected_columns, task.shard_name, csv_columns
+                    )
+                )
+            column_indices = _get_elements_indices(
```
Comment (QiJune) [inline elasticdl/python/data/reader/csv_reader.py]: Maybe `column_indices = [csv_columns.index(e) for e in selected_columns]` is shorter.

File: elasticdl/python/data/reader/csv_reader.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            for line in csv_reader:
+                line_elements = np.array(line, dtype=np.str)
+                records = line_elements[column_indices].tolist()
+                yield records
+
+    def create_shards(self):
+        pass
+
+    @property
+    def records_output_types(self):
```
Comment (QiJune) [inline elasticdl/python/data/reader/csv_reader.py:57]: It seems that the output type is not `tf.string` here.

File: elasticdl/python/tests/data_reader_test.py [hunk 5]
```diff
@@ -58,6 +58,31 @@ def test_recordio_data_reader(self):
...
+            csv_data_reader = CSVDataReader(columns=columns)
+            task = _MockedTask(0, num_records, iris_file_name)
+            records_count = 0
+            last_record = None
+            for record in csv_data_reader.read_records(task):
+                records_count += 1
+                last_record = record
+            self.assertEquals(records_count, num_records)
+            self.assertEquals(len(last_record), 5)
+
```
Comment (QiJune) [inline elasticdl/python/tests/data_reader_test.py:100]: It's better to add a unit test that generates a `tf.data.Dataset` instance with CSVDataReader.

File: elasticdl/python/data/reader/csv_reader.py (hunk 2 above)
Comment (brightcoder01) [reply elasticdl/python/data/reader/csv_reader.py]: I think for set comparision:

if not set(csv_columns) >= set(selected_columns): is not equal to 
if set(selected_columns) > set(csv_columns)

set(csv_columns) >= set(selected_columns) 
set(csv_columns) < set(selected_columns) 
can be false at the same time

File: elasticdl/python/data/reader/csv_reader.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        return tf.string
+
+    @property
+    def metadata(self):
+        return Metadata(column_names=self.selected_columns)
+
+
+def _get_elements_indices(elements, element_set):
```
Comment (brightcoder01) [inline elasticdl/python/data/reader/csv_reader.py]: elements & element_set, the parameter names are a little confusing.

File: elasticdl/python/data/reader/csv_reader.py (hunk 2 above)
Comment (QiJune) [reply elasticdl/python/data/reader/csv_reader.py]: Hmm, interesting. 
Maybe `if not set(selected_columns). issubset(set(csv_columns))` is more readable?

Comment (brightcoder01) [reply elasticdl/python/data/reader/csv_reader.py]: > csv_columns

I think so

Comment (brightcoder01) [reply elasticdl/python/data/reader/csv_reader.py]: > Hmm, interesting.
> Maybe `if not set(selected_columns). issubset(set(csv_columns))` is more readable?

I think so

Comment (workingloong) [reply elasticdl/python/data/reader/csv_reader.py]: Yes,` if not set(csv_columns) >= set(selected_columns)` is not equal to
`if set(selected_columns) > set(csv_columns)`. For example, `not set([1, 3, 4]) >= set([1, 5])` return True, but `set([1, 5]) > set([1, 3, 4])` return False. The operator "set_a >= set_b" represents that set_b is a subset of set_a.

File: elasticdl/python/data/reader/csv_reader.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    read data by line indices.
+    """
+
+    def __init__(self, **kwargs):
+        """
+        Args:
+            kwargs should contains "seq" and "columns" like
+            'seq=",",column=["sepal.length", "sepal.width", "variety"]'
+        """
+        AbstractDataReader.__init__(self, **kwargs)
```
Comment (brightcoder01) [inline elasticdl/python/data/reader/csv_reader.py:26]: Should we add check_required_kwargs here?

File: elasticdl/python/data/reader/csv_reader.py (hunk 4 above)
Comment (workingloong) [reply elasticdl/python/data/reader/csv_reader.py:57]: I have converted the outputs to string at https://github.com/sql-machine-learning/elasticdl/blob/083d2384b769124068a4f80ba51ea3b4718423b6/elasticdl/python/data/reader/csv_reader.py#L50

File: elasticdl/python/data/reader/csv_reader.py (hunk 6 above)
Comment (workingloong) [reply elasticdl/python/data/reader/csv_reader.py]: I have replaced the function with `column_indices = [csv_columns.index(e) for e in selected_columns]`

File: elasticdl/python/data/reader/csv_reader.py (hunk 7 above)
Comment (workingloong) [reply elasticdl/python/data/reader/csv_reader.py:26]: I have set the default value for arguments in `__init__`, so i think we don't need to check.

File: elasticdl/python/data/reader/csv_reader.py (hunk 2 above)
Comment (workingloong) [reply elasticdl/python/data/reader/csv_reader.py]: > Hmm, interesting.
> Maybe `if not set(selected_columns). issubset(set(csv_columns))` is more readable?

Yes!

**Review by terrytangyuan — APPROVED** [review]

File: elasticdl/python/data/reader/csv_reader.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+from elasticdl.python.data.reader.data_reader import (
+    AbstractDataReader,
+    Metadata,
+)
+
+
+class CSVDataReader(AbstractDataReader):
+    """This reader is used to read data from a csv file. It is convenient for
+    user to locally run and debug a Keras model by using this reader.
+    However, it can not be used with distribution strategy because it can not
```
Comment (terrytangyuan) [inline elasticdl/python/data/reader/csv_reader.py]: can not -> cannot

File: elasticdl/python/data/reader/csv_reader.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    read data by line indices.
+    """
+
+    def __init__(self, **kwargs):
+        """
+        Args:
+            kwargs should contains "seq" and "columns" like
```
Comment (terrytangyuan) [inline elasticdl/python/data/reader/csv_reader.py:23]: Use `check_required_kwargs` to enforce this as well

File: elasticdl/python/data/reader/csv_reader.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+            csv_columns = next(csv_reader)
+            selected_columns = (
+                csv_columns
+                if self.selected_columns is None
+                else self.selected_columns
+            )
+            if not set(selected_columns).issubset(set(csv_columns)):
+                raise ValueError(
+                    "The first line in the csv file must be column names and"
+                    "the selected columns is not in the file. The selected"
```
Comment (terrytangyuan) [inline elasticdl/python/data/reader/csv_reader.py]: is -> are

Comment (terrytangyuan) [inline elasticdl/python/data/reader/csv_reader.py]: missing space at end of each sentence strings

File: elasticdl/python/data/reader/csv_reader.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+                    "columns are {} and the columns in {} are {}".format(
+                        selected_columns, task.shard_name, csv_columns
+                    )
+                )
+            column_indices = [csv_columns.index(e) for e in selected_columns]
+            for line in csv_reader:
+                line_elements = np.array(line, dtype=np.str)
+                records = line_elements[column_indices].tolist()
```
Comment (terrytangyuan) [inline elasticdl/python/data/reader/csv_reader.py]: you can just `yield` directly without creating a new variable here

File: elasticdl/python/tests/test_utils.py [hunk 12]
```diff
@@ -155,6 +156,32 @@ def create_recordio_file(size, dataset_name, shape, temp_dir=None):
     return temp_file.name
 
 
+def create_iris_csv_file(size, columns, temp_dir=None):
+    """Creates a temporary file containing data of `recordio` format.
```
Comment (terrytangyuan) [inline elasticdl/python/tests/test_utils.py]: recordio -> csv

File: elasticdl/python/tests/test_utils.py [hunk 13]
```diff
@@ -155,6 +156,32 @@ def create_recordio_file(size, dataset_name, shape, temp_dir=None):
     return temp_file.name
 
 
+def create_iris_csv_file(size, columns, temp_dir=None):
```
Comment (terrytangyuan) [inline elasticdl/python/tests/test_utils.py:161]: missing docstring for columns

File: elasticdl/python/data/reader/csv_reader.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    """This reader is used to read data from a csv file. It is convenient for
+    user to locally run and debug a Keras model by using this reader.
+    However, it cannot be used with distribution strategy because it cannot
+    read data by line indices.
+    """
+
+    def __init__(self, **kwargs):
+        """
+        Args:
+            kwargs should contains "seq" and "columns" like
```
Comment (terrytangyuan) [inline elasticdl/python/data/reader/csv_reader.py:23]: seq -> sep. Same mistake in other places too

**Comment by terrytangyuan** [comment]
Alternative approach to consider is to support data sources from [TensorFlow I/O](https://github.com/tensorflow/io), especially since it already supports CSV along with many other formats. I created separate issue to discuss here: https://github.com/sql-machine-learning/elasticdl/issues/1585

**Review by brightcoder01 — APPROVED** [review]

**Pull request merged**